File: src/y2024/Day11.py
from functools import cache
from collections import Counter, defaultdict

def next_val(n: str):
    return ["1"] if n == "0" else [n[:len(n) // 2], n[len(n) // 2:].lstrip('0') or "0"] if len(n) % 2 == 0 else [str(int(n) * 2024)]

def map_solution(input_str: str, times: int):

    counter = Counter(input_str.split())

    for _ in range(times):
        new_counter = Counter()
        for num, count in counter.items():
            for next_num in next_val(num):
                new_counter[next_num] += count
        counter = new_counter

    return sum(counter.values())


def dfs_solution(input_str: str, times: int):
    @cache
    def dfs(n: str, depth: int = times): return 1 if depth == 0 else sum(dfs(n, depth - 1) for n in next_val(n))

    return sum(map(dfs, input_str.split()))

def read_input(filename: str):
    with open(filename, "r") as f:
        return f.read().splitlines()

if __name__ == "__main__":
    test_input = read_input("Day11_test.txt")[0]
    assert map_solution(test_input, 6) == 22
    assert dfs_solution(test_input, 6) == 22
    assert map_solution(test_input, 25) == 55312
    assert dfs_solution(test_input, 25) == 55312

    input_data = read_input("Day11.txt")[0]
    from time import time

    start = time()
    print("map= ", map_solution(input_data, 25), f"(Time: {time() - start:.4f}s)")

    start = time()
    print("dfs= ", dfs_solution(input_data, 25), f"(Time: {time() - start:.4f}s)")

    start = time()
    print("map= ", map_solution(input_data, 75), f"(Time: {time() - start:.4f}s)")

    start = time()
    print("dfs= ", dfs_solution(input_data, 75), f"(Time: {time() - start:.4f}s)")

File: src/y2024/Day11.kt
package y2024

import kotlin.time.measureTimedValue

fun main() {

    fun next(n: String) =
        when {
            n == "0" -> listOf("1")
            n.length % 2 == 0 -> listOf(
                n.take(n.length / 2),
                n.drop(n.length / 2).trimStart('0').ifEmpty { "0" })
            else -> listOf((n.toLong() * 2024).toString())
        }

    fun mapSolution(input: String, times: Int) = run {
        print("map = ")

        input.split(" ")
            .associateWith { 1L }
            .let {
                generateSequence(it) {
                    it.asSequence()
                        .flatMap { (n, v) -> next(n).map { it to v } }
                        .groupingBy { it.first }
                        .fold(0) { a, e -> a + e.second }
                }.elementAt(times).values.sum()
            }

    }

    fun dfsSolution(input: String, times: Int) = run {
        print("dfs = ")

        val map = mutableMapOf<Pair<String, Int>, Long>()

        val rec = DeepRecursiveFunction<Pair<String, Int>, Long> { (n, depth) ->
            if (depth == 0) 1
            else map.getOrPut(n to depth) { next(n).sumOf { callRecursive(it to depth - 1) } }
        }

        input.split(" ").sumOf { rec(it to times) }
    }

    val testInput = readInput("Day11_test").first()
    check(mapSolution(testInput, 6).apply { println() } == 22L)
    check(dfsSolution(testInput, 6).apply { println() } == 22L)
    check(mapSolution(testInput, 25).apply { println() } == 55312L)
    check(dfsSolution(testInput, 25).apply { println() } == 55312L)

    val input = readInput("Day11").first()
    measureTimedValue { mapSolution(input, 25) }.println()
    measureTimedValue { dfsSolution(input, 25) }.println()
    measureTimedValue { mapSolution(input, 75) }.println()
    measureTimedValue { dfsSolution(input, 75) }.println()
}